commit 6504477a1903e7ff39e7e305e6219ba89888c44b
Author: Thomas Dy <thatsmydoing@gmail.com>
Date:   Tue Dec 1 18:09:03 2020 +0900

    Use go-getter only for archive fetches

diff --git a/api/loader/getter.go b/api/loader/getter.go
index 6d397b56..d8c4e65d 100644
--- a/api/loader/getter.go
+++ b/api/loader/getter.go
@@ -71,17 +71,21 @@ func getRemoteTarget(rs *remoteTargetSpec) error {
 		log.Fatalf("Error getting wd: %s", err)
 	}
 
+	httpGetter := &getter.HttpGetter{
+		Netrc: true,
+	}
+
 	opts := []getter.ClientOption{}
 	client := &getter.Client{
 		Ctx:  context.TODO(),
 		Src:  rs.Raw,
 		Dst:  rs.Dir.String(),
 		Pwd:  pwd,
-		Mode: getter.ClientModeAny,
-		Detectors: []getter.Detector{
-			new(getter.GitHubDetector),
-			new(getter.GitDetector),
-			new(getter.BitBucketDetector),
+		Mode: getter.ClientModeDir,
+		Detectors: []getter.Detector{},
+		Getters: map[string]getter.Getter{
+			"http":  httpGetter,
+			"https": httpGetter,
 		},
 		Options: opts,
 	}
